text/0101-revamping-the-sdk-performance-api.md

```diff
@@ -0,0 +1,273 @@
+  { name: "expensiveCalc" },
+  (span: Span) => expensiveCalc()
+);
+
+// If the SDK needs async/sync typed different we can expose this
+// declare function startActiveSpanAsync<T>(
+//   spanContext: SpanContext,
+//   callback: (span: Span) => Promise<T>,
+// ): Promise<T>;
+```
+
+In the ideal case, `startActiveSpan` should generally follow this code path.
+
+1. Get the active span from the current scope
+2. If the active span is defined, create a child span of that active span based on the provided `spanContext`, otherwise create a transaction based on the provided `spanContext`.
+3. Run the provided callback
+4. Finish the child span/transaction created in step 2
+5. Remove the child span/transaction from the current scope and if it exists, set the previous active span as the active span in the current scope.
+
+If the provided callback throws an exception, the span/transaction created in step 2 should be marked as errored. This error should not be swallowed by `startActiveSpan`.
+
+`startActiveSpan` only provides the correct parent-child relationship if your platform has proper support for forking scopes. For platforms have a single hub/scope (like the mobile SDKs), this method will not lead to the correct parent-child relationship. The SDK will have to provide a different method for these platforms. One option is for `startActiveSpan` to always attach the span it creates to the root span (the transaction), which means users don't get exact parent-child relationships, but they do get relative relationships between spans using relative durations.
+
+`Sentry.startSpan` will create a span, but not set it as the active span in the current scope.
+
+```ts
+namespace Sentry {
+  declare function startSpan(spanContext: SpanContext): Span;
+}
+
+// does not get put on scope
+const span = Sentry.startSpan({ name: "expensiveCalc" });
+
+expensiveCalc();
+
+span.finish();
+```
+
+The only methods that all SDKs are required to implement are `Sentry.startActiveSpan` and `Sentry.startSpan`. For languages that need it, they add an additional method for async callbacks: `Sentry.startActiveSpanAsync`. Other languages can also attach a suffix to the methods to indicate that the spans are being started from different sources, but these are language/framework/sdk dependent.
+
+For example with go:
+
+```go
+sentry.StartSpanFromContext(ctx, spanCtx)
+```
+
+Other SDKs can also add a `startSpanFromContext` method where they can make context a variable databag that accepts headers, environment variables, etc. This is useful for easy trace propagation.
+
+Since we want to discourage accessing the transaction object directly, the `Sentry.setMeasurement` top level method will also be introduced. This will set a custom performance metric if a transaction exists. If a transaction doesn't exist, this method will do nothing. In the future, this method will attach the measurement to the span on the scope, but for now it'll only attach it to the transaction.
```

Comment:
What about `Sentry.setMetric`? `setMeasurement` might set the expectation to measure something, which is not the case.